imf/imf.py

```diff
@@ -488,15 +490,76 @@ def inverse_imf(p,
     else:
         raise NotImplementedError
 
+####This section contains the functions required to optimally sample a cluster####
+
+def prefactor(max_star,dist,mmax):
+    """
+    Returns the multiplier required for an IMF to have at most one star above m_max.
+    """
+    return 1/get_massfunc(dist).integrate(max_star,mmax)[0]
+    
+def M_cluster(m,dist,mmin,mmax):
+    """
+    Returns the mass of a cluster distributed according to some IMF where the 
+    largest star has mass m.
+    """
+    k = prefactor(m,dist,mmax)
+    return k*get_massfunc(dist).m_integrate(mmin,m)[0]+m
+
+def max_star(m,M_res,dist,mmin,mmax):
+    """
+    Returns the most massive star capable of forming in a cluster of mass M_res
+    according to the m_max/M_cluster relation. Formatted for use with root finding.
+    """
+    return M_res-M_cluster(m,dist,mmin,mmax)
+
+def approx_max_star(m,M_res):
+    """
+    Implements Eq. 10 from Pflamm-Altenburg et al. 2007 to determine the most 
+    massive star capable of forming in a cluster of mass M_res based on the 
+    m_max/M_cluster relation. Formatted for use with root finding.
+    """
+    return 2.56*np.log10(M_res)*(3.82**9.17+np.log10(M_res)**9.17)**(-1/9.17)-0.38-np.log10(m)
+
+def get_next_m(m,last_m,k,dist):
+    """
+    Returns the next smallest star in an optimally sampled cluster given the 
+    previous star and overall IMF. Formatted for use with root finding.
+    """
+    return k*get_massfunc(dist).m_integrate(m,last_m)[0]-m
+
+def opt_sample(M_res,massfunc,mmin=None,mmax=None):
+    """
+    Returns a numpy array containing stellar masses that optimally sample an
+    IMF for a cluster with mass M_res.
+    """
+    if mmin == None:
+        mmin = get_massfunc(massfunc).mmin
+    if mmax == None:
```

Comment:
should be `is None`, not `== None` .  https://stackoverflow.com/questions/26595/is-there-any-difference-between-foo-is-none-and-foo-none/26611#26611, for example